sql: Moving Resume Spans from Table Descriptors to Jobs.

m-schneider opened the pull request:
Before this change resume spans were part of table descriptors. This
created overhead because the spans were being gossiped.

After this change resume spans will be stored as part of job details.

closes #16662

Commit:
sql: Moving Resume Spans from Table Descriptors to Jobs.

Before this change resume spans were part of table descriptors. This
created overhead because the spans were being gossiped.

After this change resume spans will be stored as part of job details.

closes #16662

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17258)
<!-- Reviewable:end -->

jordanlewis commented:
Looking good! I'm quite excited about this change. I've included a bunch of stylistic comments to get the ball rolling. I think this needs a rebase as well.

---

Reviewed 19 of 19 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 221 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5c8usb_exDNgJ_Jco:-Kq5c8usb_exDNgJ_Jcp:b-z3b8uk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L221)):*
> ```Go
> 	details, ok := sc.job.Record.Details.(jobs.SchemaChangeDetails)
> 	if !ok {
> 		return errors.Errorf("wrong type of jobs")
> ```

might as well put the actual value of the details field in the error while you're here so the person who gets this error can have a chance at figuring out what the heck went wrong.

---

*[pkg/sql/backfill.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cOLf5iF6V8Ai7JeK:-Kq5cOLf5iF6V8Ai7JeL:b6pbtip) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L230)):*
> ```Go
> 		resumeSpans = append(resumeSpans, resume)
> 	}
> 	if err := txn.SetSystemConfigTrigger(); err != nil {
> ```

Pretty sure you no longer need this? Since we're not writing to the system config here anymore.

---

*[pkg/sql/backfill.go, line 375 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cX3wRvAer0fmiQCV:-Kq5cX3wRvAer0fmiQCW:b-8aqh8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L375)):*
> ```Go
> }
> 
> // Returns the jobId associated with a mutationId.
> ```

Go style says to prefix all func docs with their name - so this would be:

getJobIdForMutation returns the jobID associated with a mutationID.

Also, please rename jobId to jobID everywhere to follow the style we have of naming things ID instead of Id :)

---

*[pkg/sql/backfill.go, line 464 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cyg2PvjiH_tIPg-I:-Kq5cyg2PvjiH_tIPg-J:bt5svn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L464)):*
> ```Go
> 		job, err := sc.jobRegistry.LoadJob(ctx, jobid)
> 		if err != nil {
> 			log.Infof(ctx, "can't find job %d", jobid)
> ```

Probably want to errors.Wrapf this instead of logging.

---

*[pkg/sql/backfill.go, line 469 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cqlKaobwHBecAyym:-Kq5cqlKaobwHBecAyyn:b-1tmb68) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L469)):*
> ```Go
> 		details, ok := job.Record.Details.(jobs.SchemaChangeDetails)
> 		if !ok {
> 			return errors.Errorf("wrong type of jobs")
> ```

put the actual value in the error please

---

*[pkg/sql/schema_changer.go, line 687 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dFX9CZwhJQ2UOyzq:-Kq5dFX9CZwhJQ2UOyzr:b612jpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/schema_changer.go#L687)):*
> ```Go
> 			details, ok := sc.job.Record.Details.(jobs.SchemaChangeDetails)
> 			if !ok {
> 				return errors.Errorf("expected SchemaChangeDetails job type, got %t", reflect.TypeOf(sc.job.Record.Details))
> ```

This should be %T not %t, and you don't need to reflect.Typeof - the %T will get the type of the input value for you. See https://golang.org/pkg/fmt/

---

*[pkg/sql/distsqlrun/backfiller.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dOZP91I2fUm1CsHR:-Kq5dOZP91I2fUm1CsHS:b-3w2vwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L160)):*
> ```Go
> 	resume roachpb.Span,
> 	mutationIdx int,
> 	jobsRegistery *jobs.Registry,
> ```

typo: this should be jobsRegistery, not jobsRegistry

---

*[pkg/sql/distsqlrun/backfiller.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dYp0oBjmS8kU-DO_:-Kq5dYp0oBjmS8kU-DOa:b-jnyecc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L192)):*
> ```Go
> 		job, err := jobsRegistery.LoadJobWithTxn(ctx, jobId, txn)
> 		if err != nil {
> 			log.Infof(ctx, "can't find job %d", jobId)
> ```

errors.Wrap instead of of info

---

*[pkg/sql/distsqlrun/backfiller.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dWxhns4MOA-_qP0k:-Kq5dWxiNysUJzQxzGsJ:b-h02oe1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L197)):*
> ```Go
> 		details, ok := job.Record.Details.(jobs.SchemaChangeDetails)
> 		if !ok {
> 			return errors.Errorf("wrong type of jobs")
> ```

include the found value

---

*[pkg/sql/distsqlrun/backfiller.go, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dovrTaxaaihlqiRj:-Kq5dovrTaxaaihlqiRk:b-ylfca8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L236)):*
> ```Go
> 
> 				log.VEventf(ctx, 2, "ckpt %+v", resumeSpans)
> 				if err := txn.SetSystemConfigTrigger(); err != nil {
> ```

can get rid of this since we're not writing to the system span anymore.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eK6pdl5VduBOvk39:-Kq5eK6pdl5VduBOvk3A:bizvhyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L97)):*
> ```Go
> 	}
> 
> 	job, err := registry.LoadJob(context.TODO(), jobId)
> ```

make a ctx := context.TODO() so you don't have to keep repeating this

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5e4lmjLhbJXqvzTPe:-Kq5e4lmjLhbJXqvzTPf:b-otn0uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L101)):*
> ```Go
> 	if err != nil {
> 		log.Infof(context.TODO(), "can't find job %d", jobId)
> 		t.Fatal(err)
> ```

errors.Wrap

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5e7gX6kAu2xH8f_sZ:-Kq5e7gX6kAu2xH8f_s_:bjt05va) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L104)):*
> ```Go
> 	}
> 
> 	details, ok := job.Record.Details.(jobs.SchemaChangeDetails)
> ```

You can just cast here without the `ok` since we're in a test, if you want

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eBJB5bw1oHIwW2j7:-Kq5e0aKr4_hHWZHojPV:b-lo3mbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L106)):*
> ```Go
> 	details, ok := job.Record.Details.(jobs.SchemaChangeDetails)
> 	if !ok {
> 		t.Fatal(("wrong type of jobs"))
> ```

extra parens

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eGtAXZxSjamSZRWt:-Kq5eGtAXZxSjamSZRWu:bodjrn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L173)):*
> ```Go
> 
> 	job, _ = registry.LoadJob(context.TODO(), jobId)
> 	if err != nil {
> ```

checking the wrong err here?

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 179 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eQ-qhTsHg-riLMAr:-Kq5eQ-qhTsHg-riLMAs:b-z73yrn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L179)):*
> ```Go
> 	details, ok = job.Record.Details.(jobs.SchemaChangeDetails)
> 	if !ok {
> 		t.Fatalf("expected SchemaChangeDetails job type, got %t", reflect.TypeOf(job.Record.Details))
> ```

%T, remove the reflect.TypeOf, also again probably you can just do the cast without the ok since this is just a test.

---

*[pkg/sql/distsqlrun/server.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eZ5SPivof7YVC4F0:-Kq5eZ5SPivof7YVC4F1:b9l7o2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/server.go#L111)):*
> ```Go
> 	ClusterID uuid.UUID
> 
> 	// JobRegistry managed jobs being used by this Server.
> ```

s/managed/manages/

---

*[pkg/sql/jobs/jobs.go, line 356 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5ecwEUK_w11RUS9VU:-Kq5ecwEUK_w11RUS9VV:b-5ku6rk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/jobs/jobs.go#L356)):*
> ```Go
> 
> // SetDetails sets the details field of the currently running tracked job.
> func (j *Job) SetDetails(ctx context.Context, details interface{}) error {
> ```

Hmm I think this already exists, perhaps this PR needs a rebase!

---

*[pkg/sql/jobs/jobs.proto, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eiLGPiW6mBKWfoiX:-Kq5eiLGPiW6mBKWfoiY:b-riaz6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/jobs/jobs.proto#L37)):*
> ```protobuf
> 
> message SchemaChangeDetails {
>     int64 readAsOf = 1;
> ```

you should rebase this on top of master please!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-:-Kq5eu-kYhk4y6yjeJ5i:bdxcrji)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 15 of 19 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 221 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5c8usb_exDNgJ_Jco:-KqOCY8r35donqPYZEQr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L221)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

might as well put the actual value of the details field in the error while you're here so the person who gets this error can have a chance at figuring out what the heck went wrong.
</blockquote></details>

Done.

---

*[pkg/sql/backfill.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cOLf5iF6V8Ai7JeK:-KqOCiHpTmrXiiVya0T6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L230)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Pretty sure you no longer need this? Since we're not writing to the system config here anymore.
</blockquote></details>

Done.

---

*[pkg/sql/backfill.go, line 375 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cX3wRvAer0fmiQCV:-KqO8PljsrQwdkQtQhAt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L375)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Go style says to prefix all func docs with their name - so this would be:

getJobIdForMutation returns the jobID associated with a mutationID.

Also, please rename jobId to jobID everywhere to follow the style we have of naming things ID instead of Id :)
</blockquote></details>

Done.

---

*[pkg/sql/backfill.go, line 464 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cyg2PvjiH_tIPg-I:-KqOBLdsKHUH6h1Je_ai:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L464)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Probably want to errors.Wrapf this instead of logging.
</blockquote></details>

Done.

---

*[pkg/sql/backfill.go, line 469 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cqlKaobwHBecAyym:-KqOGlFIPVts81s90T8t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L469)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

put the actual value in the error please
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer.go, line 687 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dFX9CZwhJQ2UOyzq:-KqOHbN4Dy1H2I6vUvgF:b-opxmi6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/schema_changer.go#L687)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This should be %T not %t, and you don't need to reflect.Typeof - the %T will get the type of the input value for you. See https://golang.org/pkg/fmt/
</blockquote></details>

Ah that's really useful, thanks!

---

*[pkg/sql/distsqlrun/backfiller.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dOZP91I2fUm1CsHR:-KqOI1BalJp9K10-2L4S:b1brqle) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L160)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

typo: this should be jobsRegistery, not jobsRegistry
</blockquote></details>

You mean the other way around? :)

---

*[pkg/sql/distsqlrun/backfiller.go, line 192 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dYp0oBjmS8kU-DO_:-KqOIbLIovbyiNFDikY9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L192)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

errors.Wrap instead of of info
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller.go, line 197 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dWxhns4MOA-_qP0k:-KqOIbtrrzPHHXVa96Lv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L197)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

include the found value
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller.go, line 236 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dovrTaxaaihlqiRj:-KqOIigvPl1hkytwYa-O:b4c63be) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L236)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

can get rid of this since we're not writing to the system span anymore.
</blockquote></details>

Done. Though I kept the logging because it's useful to see what checkpoint is being written.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eK6pdl5VduBOvk39:-KqOJFMFEtVR2Ftjq0rU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L97)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

make a ctx := context.TODO() so you don't have to keep repeating this
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5e4lmjLhbJXqvzTPe:-KqOLad39QAjYQk9kwO2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L101)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

errors.Wrap
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5e7gX6kAu2xH8f_sZ:-KqOL1YgFCo0gYeOD_GT:b7s0b2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L104)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You can just cast here without the `ok` since we're in a test, if you want
</blockquote></details>

I think the error would be useful?

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 106 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eBJB5bw1oHIwW2j7:-KqOL0q9rOtJYJ46QXLX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L106)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

extra parens
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eGtAXZxSjamSZRWt:-KqOLw5ATImWBoXDLNuA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L173)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

checking the wrong err here?
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 179 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eQ-qhTsHg-riLMAr:-KqOLvUJBvvff5AZtoby:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L179)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

%T, remove the reflect.TypeOf, also again probably you can just do the cast without the ok since this is just a test.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/server.go, line 111 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eZ5SPivof7YVC4F0:-KqOM4Q7qlV7U64BKhk2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/server.go#L111)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/managed/manages/
</blockquote></details>

Done.

---

*[pkg/sql/jobs/jobs.go, line 356 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5ecwEUK_w11RUS9VU:-KqOM6Rm6rRDEb4IIlcY:b-a8gw5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/jobs/jobs.go#L356)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm I think this already exists, perhaps this PR needs a rebase!
</blockquote></details>

Yep, will rebase after replying to everything else.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 14 of 19 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 221 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5c8usb_exDNgJ_Jco:-KqOYzzA5Aikj6XHZc_w:bhju2f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L221)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`%T` will call `reflect.TypeOf` automatically. In fact, you'll get `*reflect.rtype` if you use `%T` and `reflect.TypeOf` together.

(I had to check this myself: https://play.golang.org/p/nEHv0UfK9i)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258)*
<!-- Sent from Reviewable.io -->

benesch commented:
Reviewed 14 of 19 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[pkg/sql/schema_changer.go, line 687 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dFX9CZwhJQ2UOyzq:-KqObaAHzJA49C6Ks_IW:brqrrsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/schema_changer.go#L687)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Ah that's really useful, thanks!
</blockquote></details>

Oops, didn't see this when I made the same comment above.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5e7gX6kAu2xH8f_sZ:-KqO_RjuB-yvCa9wjeRx:brc66e9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L104)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I think the error would be useful?
</blockquote></details>

The automatically-generated error if you cast without `ok` e.g.

```details := job.Record.Details.(jobs.SchemaChangeDetails)```

will be along the lines of "unable to cast to jobs.SchemaChangeDetails; interface has type FOO", so should be just as useful as the error message you've got here!
```


---

*[pkg/sql/distsqlrun/backfiller_test.go, line 173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eGtAXZxSjamSZRWt:-KqOb9aApdcuIEZJ2JTn:b-iuix5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L173)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think you still want to check the error here.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 179 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eQ-qhTsHg-riLMAr:-KqOb3xCVSqlbICVrIiH:b-5twnxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L179)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This will silently swallow type errors. I think you want `details = job.Record.Details.(jobs.SchemaChangeDetails)` (without the blank identifier).

---

*[pkg/sql/jobs/registry.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqOaRzR07sj39R2UMX8:-KqOaRzR07sj39R2UMX9:b-l0rkj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/54effcbec37735c168681902f49590dbe83f5bb4/pkg/sql/jobs/registry.go#L68)):*
> ```Go
> 	}
> 	return j, nil
> }
> ```

Ugh, it's a shame we need this. The API of `sql/jobs` isn't quite right. (My fault, not yours. This is the best we can do for now.)

---

*[pkg/sql/sqlbase/structured.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqOaIEC55XboHnFVBD5:-KqOaIEC55XboHnFVBD6:b-e7bbo3) ([raw file](https://github.com/cockroachdb/cockroach/blob/54effcbec37735c168681902f49590dbe83f5bb4/pkg/sql/sqlbase/structured.go#L33)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
> 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
> 	//"github.com/cockroachdb/cockroach/pkg/util/log"
> ```

Stray import.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Thanks! I've now rebased as well.

---

Review status: 2 of 18 files reviewed at latest revision, 21 unresolved discussions.

---

*[pkg/sql/backfill.go, line 221 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5c8usb_exDNgJ_Jco:-KqOnXDQBhoo7s5RjdcO:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L221)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

`%T` will call `reflect.TypeOf` automatically. In fact, you'll get `*reflect.rtype` if you use `%T` and `reflect.TypeOf` together.

(I had to check this myself: https://play.golang.org/p/nEHv0UfK9i)
</blockquote></details>

Thanks!

---

*[pkg/sql/schema_changer.go, line 687 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dFX9CZwhJQ2UOyzq:-KqOqfTijI3z8pzr6X7s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/schema_changer.go#L687)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Oops, didn't see this when I made the same comment above.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5e7gX6kAu2xH8f_sZ:-KqOqi5Kj108lJONXfR7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L104)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

The automatically-generated error if you cast without `ok` e.g.

```details := job.Record.Details.(jobs.SchemaChangeDetails)```

will be along the lines of "unable to cast to jobs.SchemaChangeDetails; interface has type FOO", so should be just as useful as the error message you've got here!
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 173 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eGtAXZxSjamSZRWt:-KqOqEpVUe4X4zq92jA8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L173)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think you still want to check the error here.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 179 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eQ-qhTsHg-riLMAr:-KqOnqzS8clQce2lPNfy:btjg9va) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller_test.go#L179)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

This will silently swallow type errors. I think you want `details = job.Record.Details.(jobs.SchemaChangeDetails)` (without the blank identifier).
</blockquote></details>

Right I see what you mean now. Played around with the Tour of Go site.

---

*[pkg/sql/jobs/jobs.proto, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5eiLGPiW6mBKWfoiX:-KqOsCltn9-qiCkmEOp2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/jobs/jobs.proto#L37)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

you should rebase this on top of master please!
</blockquote></details>

Done.

---

*[pkg/sql/jobs/registry.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqOaRzR07sj39R2UMX8:-KqOqLPZgYcQF6Zja-nV:bgssgxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/54effcbec37735c168681902f49590dbe83f5bb4/pkg/sql/jobs/registry.go#L68)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Ugh, it's a shame we need this. The API of `sql/jobs` isn't quite right. (My fault, not yours. This is the best we can do for now.)
</blockquote></details>

I could try to combine the two functions more?

---

*[pkg/sql/sqlbase/structured.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqOaIEC55XboHnFVBD5:-KqOqdbK6IIilwyN5fXj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/54effcbec37735c168681902f49590dbe83f5bb4/pkg/sql/sqlbase/structured.go#L33)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Stray import.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-:-KqOnIio0gvPSopsu3rE:b3mc7s2)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/backfill.go:
```diff
@@ -340,8 +343,9 @@ func (sc *SchemaChanger) getMutationToBackfill(
 	backfillType backfillType,
 	filter distsqlrun.MutationFilter,
-) (*sqlbase.DescriptorMutation, error) {
+) (*sqlbase.DescriptorMutation, int, error) {
```
can you document the integer returned?

vivekmenezes commented on pkg/sql/create.go:
```diff
@@ -185,7 +185,7 @@ func (n *createIndexNode) Start(params runParams) error {
 	}
 
-	mutationID, err := params.p.createSchemaChangeJob(params.ctx, n.tableDesc, parser.AsString(n.n))
+	mutationID, err := params.p.createSchemaChangeJob(params.ctx, n.tableDesc, parser.AsString(n.n), n.tableDesc.PrimaryIndexSpan())
```
Can we call PrimaryIndexSpan() within createSchemaChangeJob() ?

vivekmenezes commented on pkg/sql/distsqlrun/backfiller.go:
```diff
@@ -175,18 +179,38 @@ func WriteResumeSpan(
+		//This loop is finding a span in the checkpoint that fits
+		//origSpan. It then carves a spot for origSpan in the
+		//checkpoint, and replaces origSpan in the checkpoint with
+		//resume.
```
Can the code above be moved into a function?

vivekmenezes commented on pkg/sql/distsqlrun/backfiller.go:
```diff
@@ -209,18 +233,17 @@ func WriteResumeSpan(
 			}
 		}
-		// Unable to find a span containing origSpan?
+		//Unable to find a span containing origSpan?
```
nit: fix comment

vivekmenezes commented on pkg/sql/distsqlrun/backfiller_test.go:
```diff
@@ -68,14 +70,41 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
+
+	details.ResumeSpanList[0].ResumeSpans = resumeSpans
+	job.SetDetails(ctx, details)
+
```
I think this can be moved into a function and used in this test and in the main code.

vivekmenezes commented on pkg/sql/distsqlrun/backfiller_test.go:
```diff
@@ -135,7 +164,16 @@ CREATE UNIQUE INDEX vidx ON t.test (v);
+
+	details = job.Record.Details.(jobs.SchemaChangeDetails)
+
+	got := details.ResumeSpanList[0].ResumeSpans
```
same here. It will be nice to have some accessor functions that get/set the resume span on the jobs table.

vivekmenezes commented on pkg/sql/jobs/jobs.proto:
```diff
@@ -29,8 +30,19 @@ message RestoreDetails {
+  // processing can resume in the event of a node failure. The spans are
+  // non-overlapping contiguous areas of the KV space that still need to
+  // be processed.
+  repeated ResumeSpanList resume_span_list = 2 [(gogoproto.nullable) = false];
```
I believe the index here represents the index of a mutation with the same mutation id?

So a job represents many backfill operations with the same mutation id.

vivekmenezes commented on pkg/sql/schema_changer.go:
```diff
@@ -410,19 +410,19 @@ func (sc *SchemaChanger) exec(
 
 	// Find our job.
-	foundJobID := false
+	foundjobID := false
```
why change this?

vivekmenezes commented on pkg/sql/schema_changer.go:
```diff
@@ -688,7 +688,12 @@ func (sc *SchemaChanger) reverseMutations(ctx context.Context, causingError erro
+			if !ok {
+				return errors.Errorf("expected SchemaChangeDetails job type, got %T", sc.job.Record.Details)
+			}
+			details.ResumeSpanList[i].ResumeSpans = nil
```
is this written somewhere?

vivekmenezes commented on pkg/sql/sqlbase/structured.proto:
```diff
@@ -356,13 +356,6 @@ message DescriptorMutation {
-  // processing can resume in the event of a node failure. The spans are
-  // non-overlapping contiguous areas of the KV space that still need to
-  // be processed.
-  repeated roachpb.Span resume_spans = 6 [(gogoproto.nullable) = false];
```
you can add a line
```reserved 6;```

so that the tag is never used again

vivekmenezes commented on pkg/sql/table.go:
```diff
@@ -567,17 +567,26 @@ func (p *planner) getAliasedTableName(n parser.TableExpr) (*parser.TableName, er
 // MutationJob in the table descriptor.
 func (p *planner) createSchemaChangeJob(
-	ctx context.Context, tableDesc *sqlbase.TableDescriptor, stmt string,
+	ctx context.Context, tableDesc *sqlbase.TableDescriptor, stmt string, span roachpb.Span,
```
might as well compute the initialization span within this method

vivekmenezes commented on pkg/sql/table.go:
```diff
@@ -567,17 +567,26 @@ func (p *planner) getAliasedTableName(n parser.TableExpr) (*parser.TableName, er
 	}
+
+	var spanList []jobs.ResumeSpanList
+	for i := 0; i < len(tableDesc.Mutations); i++ {
```
we only want to add a ResumeSpanList for each mutation with the same mutationID

vivekmenezes reviewed (commented):
Great work on getting this to where it is. There are some minor changes I'm recommending with one major one regarding the code ignoring mutationID at one place.

jordanlewis commented:
Review status: 1 of 18 files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cOLf5iF6V8Ai7JeK:-KqOu6wqOC32L4gytLbo:b-hx1r30) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L230)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Hmm looks like this didn't get deleted in the end

---

*[pkg/sql/distsqlrun/backfiller.go, line 160 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5dOZP91I2fUm1CsHR:-KqOuHzzstMHuaRO9hd1:blyjosm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/distsqlrun/backfiller.go#L160)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

You mean the other way around? :)
</blockquote></details>

D'oh, yeah.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
> reserved 6;
Done

---

Review status: 1 of 15 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 230 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq5cOLf5iF6V8Ai7JeK:-KqUCNBUr1Y3ACkk9Qnd:bnxx08u) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d5568611b1e5f76d1721d91e47401b2b9d43c23/pkg/sql/backfill.go#L230)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm looks like this didn't get deleted in the end
</blockquote></details>

It's gone!

---

*[pkg/sql/backfill.go, line 346 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeF6-r3-346:-KqP3A-HbZCcPLDTFraV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/backfill.go#L346)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

can you document the integer returned?
</blockquote></details>

Done.

---

*[pkg/sql/schema_changer.go, line 413 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeFI-r3-413:-KqP9GpXfjfHOxbeu-Kx:b-8e6nda) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/schema_changer.go#L413)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

why change this?
</blockquote></details>

Gogland refactor gone bad. Reverted.

---

*[pkg/sql/schema_changer.go, line 696 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeFI-r3-696:-KqPGlrEIx7lVTZhwAgM:bt2vr0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/schema_changer.go#L696)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

is this written somewhere?
</blockquote></details>

Fixed!

---

*[pkg/sql/table.go, line 570 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeFM-r3-570:-KqPCmP_3BPjFQf4tFL1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/table.go#L570)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

might as well compute the initialization span within this method
</blockquote></details>

Done.

---

*[pkg/sql/table.go, line 578 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeFM-r3-578:-KqPDcv3IXfJ03MRsAqA:b-igh85r) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/table.go#L578)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

we only want to add a ResumeSpanList for each mutation with the same mutationID
</blockquote></details>

I've changed it to adding an empty list for irrelevant indices. If we add only enough of them for a mutations with the same mutation id, then on retrieval we'll have to find the first index of the mutation id and then subtract that from the actual mutation index to get the correct span. This seemed cleaner. 

---

*[pkg/sql/distsqlrun/backfiller.go, line 206 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeF8-r3-206:-KqUCRIX5k7ubMn6vnvo:b77i1l6) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/distsqlrun/backfiller.go#L206)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can the code above be moved into a function?
</blockquote></details>

I moved this and the other accessors to a function. Do you think that they're in the right place?

---

*[pkg/sql/distsqlrun/backfiller.go, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeF8-r3-244:-KqPEdXL9r9gWblozwC5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/distsqlrun/backfiller.go#L244)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

nit: fix comment
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 107 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeF9-r3-107:-KqUCY2KqwladEphT2fg:b2nreh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/distsqlrun/backfiller_test.go#L107)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think this can be moved into a function and used in this test and in the main code.
</blockquote></details>

Same as above.

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 176 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeF9-r3-176:-KqUCCETnmyG0uWnQg0h:b2nreh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/distsqlrun/backfiller_test.go#L176)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

same here. It will be nice to have some accessor functions that get/set the resume span on the jobs table.
</blockquote></details>

Same as above.

---

*[pkg/sql/jobs/jobs.proto, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeFF-r3-44:-KqP8x1_JzeD2KUPm5wB:b5zkt8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/jobs/jobs.proto#L44)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I believe the index here represents the index of a mutation with the same mutation id?

So a job represents many backfill operations with the same mutation id. 
</blockquote></details>

Yes, comment updated.

---

*[pkg/sql/create.go, line 188 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-Kq4KoVwsuE7HvHLkeF7-r3-188:-KqPAgAf5JnWmUWV5rs4:bdfcxys) ([raw file](https://github.com/cockroachdb/cockroach/blob/636de357c5915f1db62792000c0abde9367e742a/pkg/sql/create.go#L188)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can we call PrimaryIndexSpan() within createSchemaChangeJob() ?
</blockquote></details>

We're just passing the value so I don't see why not?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-:-KqP2_evY7wysqQqQQEu:b2sgkfq)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 18 files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[pkg/sql/backfill.go, line 381 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqT5XbBgKy2NK1Onuip:-KqT5XbBgKy2NK1Onuiq:beqtq2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/066cae985abb015401ac4dee6da36749b74ba065/pkg/sql/backfill.go#L381)):*
> ```Go
> }
> 
> // getjobIDForMutation returns the jobID associated with a mutationId.
> ```

getJobIDForMutation cap J

---

*[pkg/sql/backfill.go, line 413 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ3VJ8j-cTCwj0E1CI:-KqZ3VJ8j-cTCwj0E1CJ:b-58dq6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/backfill.go#L413)):*
> ```Go
> 	ctx context.Context, tableDesc *sqlbase.TableDescriptor, mutationID sqlbase.MutationID,
> ) (int64, error) {
> 	var jobID int64
> ```

No particularly strong reason for needing a jobID var

---

*[pkg/sql/distsqlrun/backfiller.go, line 181 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ5164ANui3kJdqHHm:-KqZ5164ANui3kJdqHHn:b-cn8buc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller.go#L181)):*
> ```Go
> 
> // SetResumeSpansInJob addeds a list of resume spans into a job details field.
> func SetResumeSpansInJob(
> ```

I think you can make this and the other two functions above internal by down-casing the first letter.

---

*[pkg/sql/distsqlrun/backfiller.go, line 183 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ4aoOzrAQvab_b5QZ:-KqZ4aoOzrAQvab_b5Q_:bjmko5b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller.go#L183)):*
> ```Go
> func SetResumeSpansInJob(
> 	ctx context.Context,
> 	spans *[]roachpb.Span,
> ```

Is there a reason to use a pointer here?

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 93 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ5ZlYa2Sv58GDyzRq:-KqZ5ZlZmYJV5aUABcLu:bi7yve3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller_test.go#L93)):*
> ```Go
> 			}
> 		}
> 	}
> ```

can you use the function above that maps a mutationID to a jobID?

---

*[pkg/sql/jobs/jobs.proto, line 72 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ5xmvHDEOA3kyWUHq:-KqZ5xmvHDEOA3kyWUHr:bmmtfz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/jobs/jobs.proto#L72)):*
> ```protobuf
>   // non-overlapping contiguous areas of the KV space that still need to
>   // be processed. The index represents the index of a mutation in the
>   // corresponding table descriptor muation list.
> ```

// The index represents the index of a mutation in a mutation list containing mutations for the same mutationID.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-:-KqZ6boUzPNFPU8IUUF3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Review status: 0 of 18 files reviewed at latest revision, 25 unresolved discussions, some commit checks pending.

---

*[pkg/sql/backfill.go, line 381 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqT5XbBgKy2NK1Onuip:-KqZ8PL5iL--egA8ElmE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/066cae985abb015401ac4dee6da36749b74ba065/pkg/sql/backfill.go#L381)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

getJobIDForMutation cap J
</blockquote></details>

Done.

---

*[pkg/sql/backfill.go, line 413 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ3VJ8j-cTCwj0E1CI:-KqZ8g43IDkfkiw2TUVD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/backfill.go#L413)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

No particularly strong reason for needing a jobID var
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/backfiller.go, line 181 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ5164ANui3kJdqHHm:-KqZ93O_zW32ogsXCon9:b5zibxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller.go#L181)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think you can make this and the other two functions above internal by down-casing the first letter.
</blockquote></details>

I made them external so that they could be called from backfill.go. 

---

*[pkg/sql/distsqlrun/backfiller.go, line 183 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ4aoOzrAQvab_b5QZ:-KqZ9iVtsgqC1ALZpgVt:b-xc2fgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller.go#L183)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Is there a reason to use a pointer here?
</blockquote></details>

I'm trying to avoid passing an array by value. Should I pass a slice by value instead?

---

*[pkg/sql/distsqlrun/backfiller_test.go, line 93 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ5ZlYa2Sv58GDyzRq:-KqZAf5WlcoL3ZLE4HVv:b-7hhfma) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller_test.go#L93)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

can you use the function above that maps a mutationID to a jobID?
</blockquote></details>

I don't have a SchemaChanger object to call it on.

---

*[pkg/sql/jobs/jobs.proto, line 72 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ5xmvHDEOA3kyWUHq:-KqZCFlOy-pBXuNrEHE-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/jobs/jobs.proto#L72)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

// The index represents the index of a mutation in a mutation list containing mutations for the same mutationID.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 17 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/backfiller.go, line 181 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ5164ANui3kJdqHHm:-KqZjVCS16l3wsBpl9la:b-5jhxyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller.go#L181)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I made them external so that they could be called from backfill.go. 
</blockquote></details>

gotcha! Normally we try to move such common functionality into sqlbase. No need to do that now.

---

*[pkg/sql/distsqlrun/backfiller.go, line 183 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ4aoOzrAQvab_b5QZ:-KqZi_CTu-qcMvWqkYEq:b-9cqul3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller.go#L183)):*
<details><summary><i>Previously, m-schneider (Masha Schneider) wrote…</i></summary><blockquote>

I'm trying to avoid passing an array by value. Should I pass a slice by value instead?
</blockquote></details>

Yes, passing a slice by value results in another slice that shares the same underlying array. golang encourages it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-:-KqZk4qD548ZQWb0G3oO:bnfp4nl)*
<!-- Sent from Reviewable.io -->

m-schneider commented:
Thanks!

---

Review status: 0 of 18 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlrun/backfiller.go, line 183 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-KqZ4aoOzrAQvab_b5QZ:-Kqcm5M61HlIIL88qjk3:b-4hvi0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1fa554223334f65094cfedc7465905c4488055b/pkg/sql/distsqlrun/backfiller.go#L183)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Yes, passing a slice by value results in another slice that shares the same underlying array. golang encourages it.
</blockquote></details>

Got it!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17258#-:-KqcmPUYvzg8YJ52F2Pg:b-ucfknh)*
<!-- Sent from Reviewable.io -->
